add selective simulation capabilities to TerminalComponentModeler

dmarek-flex opened the pull request:
Follows the same approach as the `ComponentModeler` where users can choose to `run_only` a subset of simulations and also may provide `element_mappings` to fill in unknown entries in the S matrix using other known entries.

The major difference is that the `coords` used in the final matrix `DataArray` are flattened so I can mix different types of Ports (Lumped and Wave), where some support modes and others do not.

<!-- greptile_comment -->

## Greptile Summary

Extends `TerminalComponentModeler` with selective simulation capabilities, enabling users to compute only specific S-matrix columns and derive missing elements through mapping relationships.

- Added `run_only` parameter to `TerminalComponentModeler` for selectively computing S-matrix columns, reducing simulation time
- Implemented `element_mappings` feature to derive unknown S-matrix elements from known entries
- Modified matrix indexing to support mixed port types (Lumped and Wave) through flattened coordinates
- Added comprehensive test coverage in `test_terminal_component_modeler.py` validating selective simulations and mappings
- Updated port naming convention in test definitions to clearly distinguish between 'coax' and 'wave' port types



<!-- /greptile_comment -->

dbochkov-flexcompute commented on tidy3d/plugins/smatrix/component_modelers/terminal.py:
```diff
@@ -148,33 +262,27 @@ def sim_dict(self) -> dict[str, Simulation]:
+        return sim_wo_source.copy(update=update_dict)
+
+    def _add_source_to_sim(
+        self, source_index: NetworkIndex, sim_wo_source: Simulation
```
is `sim_wo_source == self.base_sim` always? maybe remove the argument then?

dbochkov-flexcompute commented on tidy3d/plugins/smatrix/component_modelers/terminal.py:
```diff
@@ -95,12 +119,102 @@ def plot_sim_eps(
+        return self.matrix_indices_monitor
+
+    @cached_property
+    def matrix_indices_run_sim(self) -> tuple[NetworkIndex, ...]:
```
looks like `matrix_indices_source` and `matrix_indices_run_sim` are exactly the same as in modal component modeler. Maybe can move them to the base class?

dbochkov-flexcompute commented on tidy3d/plugins/smatrix/component_modelers/terminal.py:
```diff
@@ -208,14 +320,34 @@ def _internal_construct_smatrix(self, batch_data: BatchData) -> TerminalPortData
+            sim_data = batch_data[self._task_name(port=port, mode_index=mode_index)]
             a, b = self.compute_power_wave_amplitudes_at_each_port(port_impedances, sim_data)
-            indexer = {"f": a.f, "port_in": port_in.name, "port_out": a.port}
+            # a = a.where(a.port.isin([str(source_index)]), 0)
```
stray commented code

dbochkov-flexcompute commented on tidy3d/plugins/smatrix/component_modelers/terminal.py:
```diff
@@ -208,14 +320,34 @@ def _internal_construct_smatrix(self, batch_data: BatchData) -> TerminalPortData
+        if len(monitor_indices) == len(source_indices):
+            s_matrix = self.ab_to_s(a_matrix, b_matrix)
+        else:
+            a_diag = np.diagonal(a_matrix, axis1=1, axis2=2)
+            # Scale each column by the corresponding diagonal entry
+            s_matrix = b_matrix / a_diag[:, np.newaxis, :]
```
is this an approximation to s-matrix? or is it still exact one? If it is an approximation we should probably show a warning

dmarek-flex replied:
Note sure what the best approach is here. In an ideal simulation, the two techniques would be equivalent because the `a_matrix` would be a diagonal matrix. In practice though, even if you only excite one Port (say a `ModeSource`) there might be reflections from PMLs or other affects that make nonzero non-diagonal entries enter into the `a_matrix` at the ports that are turned off. The `ComponentModeler` takes the approach that assumes the off-diagonal entries are 0, and I believe it is the only way for calculating partial S matrices.

I'm leaning towards making the default in `TerminalComponentModeler` to be the same as `ComponentModeler`, so just taking the diagonal of `a_matrix`. But I might run some tests, to see how S parameters change when using this approach.

dbochkov-flexcompute commented on tidy3d/plugins/smatrix/component_modelers/terminal.py:
```diff
@@ -208,14 +320,34 @@ def _internal_construct_smatrix(self, batch_data: BatchData) -> TerminalPortData
+            # Scale each column by the corresponding diagonal entry
+            s_matrix = b_matrix / a_diag[:, np.newaxis, :]
+
+        # element can be determined by user-defined mapping
+        for (row_in, col_in), (row_out, col_out), mult_by in self.element_mappings:
+            coords_from = {
+                "port_in": col_in,
+                "port_out": row_in,
+            }
+            coords_to = {
+                "port_in": col_out,
+                "port_out": row_out,
+            }
+            s_matrix.loc[coords_to] = mult_by * s_matrix.loc[coords_from].values
+
```
I am not an expert in this, so maybe that's why I don't see this immediately, but do we get identical s-matrices if we run a component modeler directly without any mappings and with appropriate mappings? I could see that if element mappings were used to populate a and b until they are full, and then the s-matrix were computed as usual

dmarek-flex replied:
The element mappings are when the user has some additional knowledge of the component, like that it is symmetric so that port 1 and port 2 are identical. You could also assume that the system is reciprocal which forces the s matrix itself to be symmetric. 

So maybe theoretically you should get identical s matrices, but in practice there will be some numerical error that might change the values.

weiliangjin2021 reviewed (approved):
All look good to me.

dmarek-flex commented:
Going to be conflicts with https://github.com/flexcompute/tidy3d/pull/2663, so I might wait for that to merge first.

FYI @daquinteroflex

daquinteroflex commented:
Actually I'd argue it's probably better to merge this if it's ready and continue the refactor using this

dmarek-flex committed:
add selective simulation capabilities to TerminalComponentModeler

refactored common code into AbstractComponentModeler

add toggle for controlling how s matrix is calculated

added validation of run_only and element_mappings for modelers
